pkg/gitrepository/manager.go

```diff
@@ -0,0 +1,139 @@
+//counterfeiter:generate -o fakes/fake_kubernetes.go . Kubernetes
+//Kubernetes interface for itneracting with kubernetes
+type Kubernetes interface {
+	Get(ctx context.Context, key client.ObjectKey, obj client.Object) error
+	Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error
+	Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error
+}
+
+//NewManager returns a Manager struct
+func NewManager(ctx context.Context, namespace string, kClient Kubernetes, timeout, interval time.Duration) *Manager {
+	return &Manager{
+		kClient:   kClient,
+		namespace: namespace,
+		ctx:       context.TODO(),
+		timeout:   timeout,
+		interval:  interval,
+	}
+}
+
+//CreateAndWaitForResources creates the gitrepository resources and waits for them to be created
+func (m *Manager) CreateAndWaitForResources(r profilesv1.Repository, instances []Instance) ([]*sourcev1.GitRepository, error) {
+	var gitResources []*sourcev1.GitRepository
+	for _, instance := range instances {
+		gitRes := makeGitRepository(r, instance.Tag, instance.Path, m.namespace)
+		err := m.kClient.Create(m.ctx, gitRes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create gitrepository: %w", err)
+		}
+		gitResources = append(gitResources, gitRes)
+	}
+
+	for _, gitRes := range gitResources {
+		err := m.waitForURL(gitRes)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return gitResources, nil
+}
+
+func (m *Manager) waitForURL(gitRes *sourcev1.GitRepository) error {
+	startTime := time.Now()
+	for {
+		err := m.kClient.Get(m.ctx, client.ObjectKeyFromObject(gitRes), gitRes)
+		if err != nil {
+			return fmt.Errorf("failed to get gitrepository: %w", err)
+		}
+		if gitRes.Status.URL != "" {
+			return nil
+		}
+
+		if time.Since(startTime) > m.timeout {
+			return fmt.Errorf("timed out waiting for %s/%s gitrepository.Status.URL to be populated", gitRes.Namespace, gitRes.Name)
+		}
+		time.Sleep(m.interval)
+	}
+}
+
+func makeGitRepository(r profilesv1.Repository, tag, path, namespace string) *sourcev1.GitRepository {
+	ignore := fmt.Sprintf(`# exclude all
+/*
+# include deploy dir
+!/%s`, path)
+
+	repo := &sourcev1.GitRepository{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      makeGitRepoName(tag, r.URL),
+			Namespace: namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       sourcev1.GitRepositoryKind,
+			APIVersion: sourcev1.GroupVersion.String(),
+		},
+		Spec: sourcev1.GitRepositorySpec{
+			URL: r.URL,
+			Reference: &sourcev1.GitRepositoryRef{
+				Tag: tag,
+			},
+			Ignore: &ignore,
+		},
+	}
+
+	if r.SecretRef != nil {
+		repo.Spec.SecretRef = r.SecretRef
+	}
+	return repo
+}
+
+func makeGitRepoName(tag, url string) string {
+	urlParts := strings.Split(url, "/")
+	repo := strings.TrimRight(urlParts[len(urlParts)-1], ".git")
+
+	splitTag := strings.Split(tag, "/")
+	if len(splitTag) == 2 {
+		return fmt.Sprintf("%s-%s-%s", repo, splitTag[0], splitTag[1])
+	}
+	return fmt.Sprintf("%s-%s", repo, tag)
```

Comment:
We probably have to sanitise tag here, because it doesn't use domain naming conventions, right?
Like we did with GitRepository object in PCTL using sanitise. It was just doing this:

```go
var invalid = regexp.MustCompile(`[^a-zA-Z0-9_-]+`)

// SanitiseBranchName makes the branch domain name friendly.
func SanitiseBranchName(branch string) string {
	sanitized := invalid.ReplaceAllString(branch, "")
	return strings.ReplaceAll(sanitized, "_", "-")
}
```

Reply:
https://git-scm.com/docs/git-check-ref-format provides a fair bit of detail on what Git considers to be a valid branch name...and I'm pretty confident that GitHub accepts everything that Git does :-)

Reply:
The error doesn't come from git. If we include in the kubernetes object name anything that isn't DNS compatible, like _ it will throw an error.

Reply:
I think since this is using the tag format (string/semver) not an arbitrary branch name it should be okay?

Reply:
Yeah, sure. If you think it couldn't happen, I'm happy to ignore. :)